### DISCUSSION TITLE
Support for wxWidgets 3.1.6 (in particular wxBitmapBundle)

### DISCUSSION THREAD

**Discussion by jmoraleda** [wxFormBuilder/wxFormBuilder#739] [Q&A]
Hello all,

Today I updated my wxWidgets system to 3.1.6 and not surprisingly the wxFormBuilder generated code has stopped working. From a quick glance, I think most of the problems arise from places where we pass `wxBitmap` but the new code expects `wxBitmapBundle`. 

Is there already an effort currently underway to support wxBitmapBundle? If not, would it be reasonable to create a new branch to support it? 

I am going to look into this and I'll report my findings in this discussion.

**Comment by sodevel** [comment]
Actually, this is surprising because i am using wxWidgets master snapshots for quite some time now and haven't discovered any issues so far. In general, the `wxBitmapBundle` update should be pretty well backward compatible, at least from the calling side, you should still be able to pass in a plain `wxBitmap`. The situation is a bit different if you override wxWidgets methods, in that case you need to adapt to new signatures.

So far, there was one location in the wxFB source that needed to be adjusted, i don't know which generated code might be affected as well. Even in my biggest projects, all generated code compiles fine, but i am using XRC type resources only.

If generated code does not compile anymore, this should be fixed, but this must not break wxWidgets < 3.1.6 versions, these versions are just way too common these days. Supporting the new features of `wxBitmapBundle` (i'm not too familiar with these right now, i'm thinking about adding multiple bitmaps or even svg images) should be very hard, the wxFB data model doesn't support this, the way the `wxPropertyGrid` is used right now should be plain wrong, i would refrain from doing this now.

**Comment by jmoraleda** [comment]
I use wxFormBuilder to generate Python code. After the recent update of wxPython to wxWidgets 3.1.6 some of the generated code does not work any more because passing `Bitmaps` to `Buttons` (and a few other widgets) is no longer supported, at least in wxPython, and now one needs to pass `BitmapBundles`.

I also don't see a clear path for wxFormBuilder to support bundles with multiple bitmaps and agree that the only reasonable way in the short and medium term is for wxFormBuilder to support _single bitmap bundles_. Fortunately that is straight forward using `BitmapBundle::FromBitmap` which was included exactly for this purpose.

 I made a number of small changes to make wxFormBuilder generate Python code that is compatible with the latest wxPython. You can see those changes at https://github.com/jmoraleda/wxFormBuilder/commit/ec4f0e47dd27a6eb960915e71a173c413a9c1fc0 

Unfortunately the new generated code is not backward compatible and I don't see a way to make it so.

I assume that similar changes will be required for the other languages, but I am not familiar with their wxWidgets support, so perhaps passing Bitmap is still supported in their bindings but since the C++ API has not gone that way, I am doubtful.

**Comment by sodevel** [comment]
Sorry, maybe i should clarify i'm a C++ guy, i can speak for the C++ side only.

A quick check shows that on the C++ side (almost?) everything that used to accept a `wxBitmap` now requires a `wxBitmapBundle`. However, `wxBitmapBundle` can be automatically created from a `wxBitmap`, thats why my code still compiles fine.

I know almost nothing about Python, the latest wxPython is from 2020 and the documentation on the website is from the same date, changes for wxWidgets 3.1.6 have been made only recently to the project. I would guess for wxPython this is still a work in progress and this automatic conversion is just missing currently, unless such automatic conversion is technically not possible.

I would ask the wxPython project first if they plan on adding such feature before making changes to wxFormBuilder. For C++ nothing needs to be done, i think PHP and Lua can be safely ignored, these ports look pretty dead to me.

**Reply by jmoraleda** [reply]
Thank you. wxPython is under very active development. The latest commit was only a few hours ago: https://github.com/wxWidgets/Phoenix wxPython switches to the latest release of wxWidgets within a few weeks of it being made. As of a few days ago it is using `3.1.6` hence me bringing up this discussion. I build wxPython from source but one can get prebuilt packages at https://wxpython.org/Phoenix/snapshot-builds/?C=M;O=D

I believe wxFormBuilder definitely should continue supporting the most recent wxPython actively. I use wxPython with and without wxFormBuilder and I am happy to help maintain wxFormBuilder to support it.

Whether wxPython should automatically cast bitmaps, icons and images to bitmap bundles like C++ apparently does, or whether we should change wxFormBuilder python code generation to explicitly call wx.FromBitmap as needed is a good question. I just raised it within wxPython as well per your suggestion: https://github.com/wxWidgets/Phoenix/issues/2153

Let's see what the arguments are for both options.

**Reply by jmoraleda** [reply]
I believe that after this commit in wxPython https://github.com/wxWidgets/Phoenix/commit/c054d18112c5fc7e5bee91a76a50fc1172768f9b this no longer an issue as wxPython now also automatically converts to `Bitmap`, `Icon` and `Image` to `BitmapBundle`
